pkg/api/types.go

```diff
@@ -180,6 +180,112 @@ type VolumeSource struct {
 	GitRepo *GitRepo `json:"gitRepo"`
 	// Secret represents a secret that should populate this volume.
 	Secret *SecretSource `json:"secret"`
+
+	// TODO (per thockin) https://github.com/GoogleCloudPlatform/kubernetes/pull/4055#discussion_r24865963
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimAttachment *PersistentVolumeClaimAttachment `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:",inline"`
```

Comment:
Brian's comment is mostly what (I think) I said too - don't repurpose VolumeSource.  I wouldn't even bother inlining it - just define a new struct, maybe PersistentVolumeSource, that overlaps with VolumeSource.  every time someone wants to add a new volume type, they need to decide "is this available directly to pods, available persistently, or both?"

If/when we get to a place that has all volumes available in both modes, we can de-dup.